iwyu.cc

```diff
@@ -2635,12 +2601,46 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return visitor_state_->preprocessor_info;
   }
 
-  bool IsProvidedTypeComponent(const ASTNode* node) const {
-    return node->HasAncestorOfType<TypedefNameDecl>() &&
-           !CodeAuthorWantsJustAForwardDeclare(node->GetAs<Type>(),
-                                               node->GetLocation());
+  pair<bool, const char*> CanBeProvidedTypeComponent(
+      const ASTNode* node) const {
+    if (node->HasAncestorOfType<TypedefNameDecl>())
+      return pair(true, nullptr);
+
+    if (const FunctionDecl* decl = node->GetAncestorAs<FunctionDecl>()) {
+      // No point in author-intent analysis of function definitions
+      // in source files, or for builtins, or for friend declarations.
+      if (!IsInHeader(decl) || IsFriendDecl(decl))
+        return pair(false, nullptr);
+      const Type* return_type = decl->getReturnType().getTypePtr();
+      if (node->StackContainsContent(return_type)) {
+        const bool can_be_provided =
+            !decl->isThisDeclarationADefinition() &&
+            !IsPointerOrReferenceAsWritten(return_type);
+        return pair(can_be_provided,
+                    can_be_provided ? "(for fn return type)" : nullptr);
+      }
```

Comment:
Isn't this just
```
if (!decl->isThisDeclarationADefinition() &&
    !IsPointerOrReferenceAsWritten(return_type)) {
  return pair(true, "(for fn return type)");
}
```
... and otherwise it falls through to the default return? Or does the line get too long so this formats better?

Reply:
Correct, thanks!